### PR TITLE
Implements Isolation Forest Feature Request

### LINKED ISSUE #195: Feature Request: IsolationForest
I'd really like to see the IsolationForest in golearn. The IsolationForest is a outlier detection algorithm as described in [Isolation Forest (Fei Tony Liu, Kai Ming Ting, Zhi-Hua Zhou)](https://cs.nju.edu.cn/zhouzh/zhouzh.files/publication/icdm08b.pdf). It is also supported by pythons scikit-learn ([IsolationForest in scikit-learn](http://scikit-learn.org/stable/modules/generated/sklearn.ensemble.IsolationForest.html#sklearn.ensemble.IsolationForest))

### PR DESCRIPTION
**Pull request by Yushgoel**
Implements Isolation Forest for Feature Request #195 .

Uses the CART regression structs for implementing Isolation Forest.
Returns anomaly score for each data point as a prediction. 

Closes #195

### CODE REVIEW

File: trees/isolation.go [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+		upperNode.Right = &rightNode
+	}
+
+	return upperNode
+}
+
+// Get a random subset of the data. Helps making each tree in forest different.
+func getRandomData(data [][]float64, subSpace int) [][]float64 {
+	var randomData [][]float64
+	rand.Seed(time.Now().UnixNano())
```
Comment (Sentimentron) [inline trees/isolation.go:125]: See above comment about seeding.

File: trees/isolation.go [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+type IsolationForest struct {
+	nTrees   int
+	maxDepth int
+	subSpace int
+	trees    []regressorNode
+}
+
+// Select A random feature for splitting from the data.
+func selectFeature(data [][]float64) int64 {
+	rand.Seed(time.Now().UnixNano())
```
Comment (Sentimentron) [inline trees/isolation.go:20]: Re-seeding the random number generator will make things harder to reproduce exactly. It may be best to leave this to the program using the library.

**Review by Sentimentron — COMMENTED** [review]
Cool, thanks for this! Couple of things:
* Did you generate the `gaussian_outliers.csv` file yourself or did it come from somewhere else?
* rand.Seed() makes things hard to test, precisely because the result is not the same each time. It looks like in Go 1.13, the order of random numbers could be a bit different (which might be why the test's failing).
* It might be good to add the right methods for saving and loading the random forest (but how this works is not very well-documented, so shouldn't be a blocker IMO).

File: trees/isolation.go (hunk 2 above)
Comment (Yushgoel) [reply trees/isolation.go:20]: Hi,

So the gaussian_outliers.csv was generated by me. It is an example of a standard type of dataset used for outlier detection. It essentially has a lot of data points from gaussian data (which are normal), and then we have 10 outliers I added.

I think the idea behind an isolation forest is that everything is as random as possible. Only when this condition is met, outliers are filtered out early. That's why I changed  the rand.seed, because otherwise the seed is fixed and random numbers are totally predictable. 

So then I'll remove the lines that change the seed and add another commit?

**Review by Sentimentron — APPROVED** [review]
Cool, awesome, thanks for this!
